<div class="alert alert-block alert-info">
<b>Question 1: </b> Given some sample data, write a program to answer the following: <a href="https://docs.google.com/spreadsheets/d/16i38oonuX1y1g7C_UAmiK9GkY7cS-64DfiDMNiR41LM/edit#gid=0"> Click here to access the required data set
    </a>
</div>

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 

- Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 

- What metric would you report for this dataset?
- What is its value?


<div class="alert alert-success" role="alert">
    <b> Answer 1: </b> 
    
-There are outliers in the dataset. Therefore, we are getting extreme AOV when it is calculated with Mean since Mean is sensitive to outliers.

- I have used 2 approches to get the average order value. 
    1. Using the Median value instead of Mean value.
    2. Removing all the outliers from the dataset and re-calculate the Mean.
    
    
- The average order value (AOV) calculated in the original dataset using Median: \$284

- The average order value (AOV) calculated with the Mean after all outliers has been removed: \$294
    

Please see below for the detailed calculation.
</div>

In [51]:
import pandas as pd

Loading the dataset

In [52]:
data = pd.read_csv("2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv")

Reading the first 10 rows of the dataset

In [53]:
data.head(10)

,order_id,shop_id,user_id,order_amount,total_items,payment_method,created_at
0,1,53,746,224,2,cash,2017-03-13 12:36:56
1,2,92,925,90,1,cash,2017-03-03 17:38:52
2,3,44,861,144,1,cash,2017-03-14 4:23:56
3,4,18,935,156,1,credit_card,2017-03-26 12:43:37
4,5,18,883,156,1,credit_card,2017-03-01 4:35:11
5,6,58,882,138,1,credit_card,2017-03-14 15:25:01
6,7,87,915,149,1,cash,2017-03-01 21:37:57
7,8,22,761,292,2,cash,2017-03-08 2:05:38
8,9,64,914,266,2,debit,2017-03-17 20:56:50
9,10,52,788,146,1,credit_card,2017-03-30 21:08:26


In [54]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   order_id        5000 non-null   int64 
 1   shop_id         5000 non-null   int64 
 2   user_id         5000 non-null   int64 
 3   order_amount    5000 non-null   int64 
 4   total_items     5000 non-null   int64 
 5   payment_method  5000 non-null   object
 6   created_at      5000 non-null   object
dtypes: int64(5), object(2)
memory usage: 273.6+ KB


Takeaway so far: There are 5000 rows of data

In [55]:
print('There are',data.duplicated().sum(), 'duplicate values in dataset')

There are 0 duplicate values in dataset


**Statistical Summary of the data**

In [56]:
data.describe()

,order_id,shop_id,user_id,order_amount,total_items
count,5000.000000,5000.000000,5000.000000,5000.000000,5000.00000
mean,2500.500000,50.078800,849.092400,3145.128000,8.78720
std,1443.520003,29.006118,87.798982,41282.539349,116.32032
min,1.000000,1.000000,607.000000,90.000000,1.00000
25%,1250.750000,24.000000,775.000000,163.000000,1.00000
50%,2500.500000,50.000000,849.000000,284.000000,2.00000
75%,3750.250000,75.000000,925.000000,390.000000,3.00000
max,5000.000000,100.000000,999.000000,704000.000000,2000.00000


- In average number of sneakers ordered in a month is 9 but average the order amount is \$3145 which we already knew and that seems too high for affordable sneakers.


- When looking in the median data, the order amount is actually more than 10 times lower than average value(\$284). 


- The median order amount tells us that in the last 30 days, 50% of order amount was \$284 and below. 


- Also 75% of order amount was around \$390 and the total items were only 3.


- Looking into the maximum value, we can tell the highest order amount of the month was \$704k and total items ordered were 2000, which seem pretty high.This is the outlier in our data which is giving the high average order amount.


- When we have outliers, one way to calculate average is using the median since Mean is sensitive to outliers. That gives us \$284 average order value.

**Another way is excluding the outliers from the dataset and re-calculate the average. Let's see what we would get:**

Function to find the outliers in dataset using interquartile range

In [57]:
def find_outliers_IQR(data):

   q1=data.quantile(0.25)

   q3=data.quantile(0.75)

   IQR=q3-q1

   outliers = data[((data<(q1-1.5*IQR)) | (data>(q3+1.5*IQR)))]

   return outliers

Outliers in Order_amount

In [58]:
outliers = find_outliers_IQR(data['order_amount'])

print('number of outliers:' + str(len(outliers)))

print('max outlier value:' + str(outliers.max()))

print('min outlier value: '+ str(outliers.min()))

number of outliers:141
max outlier value:704000
min outlier value: 735


Outliers in total_items

In [59]:
outliers = find_outliers_IQR(data['total_items'])

print('number of outliers:' + str(len(outliers)))

print('max outlier value:' + str(outliers.max()))

print('min outlier value: '+ str(outliers.min()))

number of outliers:18
max outlier value:2000
min outlier value: 8


Now, I am excluding the items above 8 and order amount above 735 and recalculate the statistical summary

In [60]:
Data_withoutOutliers= data.query('order_amount < 735' or 'total_items < 8')

In [61]:
Data_withoutOutliers.head()

,order_id,shop_id,user_id,order_amount,total_items,payment_method,created_at
0,1,53,746,224,2,cash,2017-03-13 12:36:56
1,2,92,925,90,1,cash,2017-03-03 17:38:52
2,3,44,861,144,1,cash,2017-03-14 4:23:56
3,4,18,935,156,1,credit_card,2017-03-26 12:43:37
4,5,18,883,156,1,credit_card,2017-03-01 4:35:11


In [62]:
Data_withoutOutliers.describe()

,order_id,shop_id,user_id,order_amount,total_items
count,4859.000000,4859.000000,4859.000000,4859.000000,4859.000000
mean,2497.395966,49.852645,849.905742,293.715374,1.950196
std,1443.356555,29.049171,86.887496,144.453395,0.919791
min,1.000000,1.000000,700.000000,90.000000,1.000000
25%,1244.500000,24.000000,776.000000,162.000000,1.000000
50%,2498.000000,50.000000,850.000000,280.000000,2.000000
75%,3749.500000,74.000000,925.000000,380.000000,3.000000
max,5000.000000,100.000000,999.000000,730.000000,5.000000


Here, we got the average order_amount $294 after removing the outliers.

To conclude, for calculating the average order value, we can either use Median value or average value after the outliers has been removed.

<div class="alert alert-block alert-info">
<b>Question 2: </b> For this question you’ll need to use SQL.
<a href="https://www.w3schools.com/SQL/TRYSQL.ASP?FILENAME=TRYSQL_SELECT_ALL"> Follow this link </a> to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below. 
</div>

a. How many orders were shipped by Speedy Express in total?

<div class="alert alert-success" role="alert">
    <b> Answer a: </b> 
54 orders were shipped by Speedy Express in total.
    
SQL Code below:

In [67]:
%%sql

SELECT 
    ShipperName,
    COUNT(Orders.OrderID) as Total
FROM 
    Orders
    INNER JOIN Shippers on Orders.ShipperID = Shippers.ShipperID
    
WHERE Shippers.ShipperName is "Speedy Express";

UsageError: Cell magic `%%sql` not found.


b. What is the last name of the employee with the most orders?

<div class="alert alert-success" role="alert">
    <b> Answer b: </b> 
"Peacock" is the last name of the employee with the most orders, which total of 40 orders.
    
SQL Code below:

In [68]:
%%sql
SELECT 
	Employees.LastName,
	COUNT(Orders.OrderID)
    
FROM 
	Employees
    INNER JOIN Orders ON Orders.EmployeeID = Employees.EmployeeID
 GROUP BY
 Employees.LastName

UsageError: Cell magic `%%sql` not found.


c. What product was ordered the most by customers in Germany?

<div class="alert alert-success" role="alert">
    <b> Answer c: </b> 
"Steeleye Stout" was ordered the most by customers in Germany (100 quantity).
    
SQL Code below:

In [1]:
%%sql
SELECT 
	OrderDetails.quantity,
    products.ProductName
    FROM Customers
    JOIN Orders ON Orders.CustomerID = Customers.CustomerID
    JOIN OrderDetails ON OrderDetails.OrderID = Orders.OrderID
    JOIN products ON products.ProductID = OrderDetails.ProductID
    

    WHERE
		Customers.Country is "Germany"
GROUP BY products.ProductName
ORDER BY OrderDetails.quantity DESC;

UsageError: Cell magic `%%sql` not found.
